File: questions/155.MinStack.ts
class MinStack {
  stack: number[] = []
  minStack: number[] = []

  push(val: number): void {
    this.stack.push(val)

    const minStackValue = this.minStack.length
      ? Math.min(this.minStack[this.minStack.length - 1], val)
      : val
    this.minStack.push(minStackValue)
  }

  pop(): void {
    this.stack.pop()
    this.minStack.pop()
  }

  top(): number {
    return this.stack[this.stack.length - 1] ?? 0
  }

  getMin(): number {
    return this.minStack[this.minStack.length - 1]
  }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */
const stack = new MinStack()
stack.push(-2)
stack.push(0)
stack.push(-3)
console.log(stack.getMin())
stack.pop()
console.log(stack.top())
console.log(stack.getMin())

File: leetcode/155.MinStack.go
package leetcode

import "math"

type MinStack struct {
	Stack    []int
	MinStack []int
}

func (m *MinStack) Push(val int) {
	m.Stack = append(m.Stack, val)

	minStackValue := val
	if len(m.MinStack) > 0 {
		minStackValue = int(math.Min(float64(m.MinStack[len(m.MinStack)-1]), float64(val)))
	}

	m.MinStack = append(m.MinStack, minStackValue)
}

func (m *MinStack) Pop() {
	m.Stack = m.Stack[:len(m.Stack)-1]
	m.MinStack = m.MinStack[:len(m.MinStack)-1]
}

func (m *MinStack) Top() int {
	if len(m.Stack) == 0 {
		return 0
	}

	return m.Stack[len(m.MinStack)-1]
}

func (m *MinStack) GetMin() int {
	if len(m.MinStack) == 0 {
		return 0
	}

	return m.MinStack[len(m.MinStack)-1]
}
